File: backend/arceus/core/delegation_memory.py
"""
DelegationMemoryManager — Arceus adapter for Employee delegation memory flow.

Spec reference: hippocampus_design_v6.md section 8.3 / Flow B
Key principle: memories are COPIED, never referenced.
"""
from __future__ import annotations

from arceus.core.hippocampus.hippocampus import Hippocampus
from arceus.core.hippocampus.types import MemoryType, MemoryUnit
from arceus.core.memory_scope import ArceusMemoryScope


class DelegationMemoryManager:
    """
    Handles memory context injection during Employee-to-Employee delegation.
    Knows about delegation authority and org hierarchy (Arceus-domain logic).
    """

    def __init__(self, scope: ArceusMemoryScope | None = None) -> None:
        self._scope = scope or ArceusMemoryScope()

    async def prepare_delegation_context(
        self,
        from_hippocampus: Hippocampus,
        to_hippocampus: Hippocampus,
        from_agent_id: str,
        to_agent_id: str,
        startup_id: str,
        task_id: str,
        task_description: str,
        top_k: int = 10,
    ) -> list[MemoryUnit]:
        """
        Query delegator's hippocampus for task-relevant memories,
        COPY them into the delegatee's task-scoped container.
        Returns the list of copied memories.
        """
        from_container = self._scope.employee_container(startup_id, from_agent_id)
        task_container = self._scope.task_container(startup_id, task_id)

        relevant = await from_hippocampus.recall(
            query=task_description,
            container=from_container,
            top_k=top_k,
        )

        copied: list[MemoryUnit] = []
        for mem in relevant:
            if not isinstance(mem, MemoryUnit):
                continue
            copy = MemoryUnit(
                agent_id=to_agent_id,
                startup_id=startup_id,
                content=mem.content,
                embedding=mem.embedding,
                memory_type=MemoryType.DYNAMIC,
                confidence=mem.confidence,
                container=task_container,
                visibility=mem.visibility,
                source_type="delegation",
                source_id=from_agent_id,
                provenance=f"Delegated from {from_agent_id}",
                metadata={"delegated_from": from_agent_id},
            )
            await to_hippocampus._vector_store.upsert(copy)
            copied.append(copy)

        return copied

    async def internalize_delegation_result(
        self,
        delegator_hippocampus: Hippocampus,
        delegator_agent_id: str,
        startup_id: str,
        learnings: list[str],
        quality: float,
    ) -> None:
        """
        After delegation completes, internalize verified learnings
        into the delegator's personal memory.
        Quality >= 0.6 = dynamic, >= 0.9 = static.
        """
        if quality < 0.6:
            return

        container = self._scope.employee_container(startup_id, delegator_agent_id)
        for learning in learnings:
            memory_type = MemoryType.STATIC if quality >= 0.9 else MemoryType.DYNAMIC
            await delegator_hippocampus.remember(
                content=learning,
                container=container,
                memory_type=memory_type,
            )

File: server/src/services/delegation-memory.ts
import type { HippocampusBridge, MemoryListItem } from "./hippocampus-contract.js";
import { MemoryContainers } from "./memory-scope.js";

export interface DelegationResult {
  copiedCount: number;
  failedCount: number;
  memories: MemoryListItem[];
}

export class DelegationMemoryService {
  constructor(private readonly bridge: HippocampusBridge) {}

  async prepareDelegationContext(
    fromAgentId: string,
    toAgentId: string,
    startupId: string,
    taskId: string,
    taskDescription: string,
    topK = 10,
  ): Promise<DelegationResult> {
    const fromContainer = MemoryContainers.employee(startupId, fromAgentId);
    const relevant = await this.bridge.recall(
      fromAgentId,
      taskDescription,
      fromContainer,
      topK,
    );
    const taskContainer = MemoryContainers.task(startupId, taskId);

    const copyOps = relevant.items.map((memory) =>
      this.bridge.remember(
        toAgentId,
        `[delegated:${fromAgentId}] ${memory.content}`,
        taskContainer,
        "dynamic",
      ).then((result) => ({
        id: result.id,
        content: result.content,
        memory_type: result.memory_type,
        confidence: result.confidence,
        relevance_score: memory.relevance_score ?? 0,
        container: taskContainer,
        visibility: null,
        created_at: null,
        updated_at: null,
        access_count: 0,
      })),
    );

    const settled = await Promise.allSettled(copyOps);
    const fulfilled = settled.filter(
      (
        result,
      ): result is PromiseFulfilledResult<Awaited<(typeof copyOps)[number]>> => result.status === "fulfilled",
    );
    const memories: MemoryListItem[] = fulfilled.map((result) => result.value);

    return {
      copiedCount: memories.length,
      failedCount: settled.filter((result) => result.status === "rejected").length,
      memories,
    };
  }

  async internalizeDelegationResult(
    agentId: string,
    startupId: string,
    learnings: string[],
    quality: number,
  ): Promise<{ internalized: number }> {
    if (quality < 0.6) {
      return { internalized: 0 };
    }

    const container = MemoryContainers.employee(startupId, agentId);
    const memoryType = quality >= 0.9 ? "static" : "dynamic";
    const settled = await Promise.allSettled(
      learnings.map((learning) => this.bridge.remember(agentId, learning, container, memoryType)),
    );

    return {
      internalized: settled.filter((result) => result.status === "fulfilled").length,
    };
  }
}
